tfx/experimental/distributed_inference/graphdef_experiments/subgraph_partitioning/beam_pipeline.py

```diff
@@ -0,0 +1,339 @@
+# Copyright 2020 Google LLC. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Batch inference with Beam on partitioned subgraphs.
+
+There are two libraries representing two stages: graph_partition and
+beam_pipeline. After graph_partition produces lists of ExecutionSpecs that
+contain partitioned subgraphs, beam_pipeline constructs a Beam pipeline that
+executes the partitioned subgraphs. The results from the Beam pipeline should
+be the same as the results from the original model.
+
+  Typical usage example:
+  ```
+  # Obtained graph_name_to_specs from the graph partition
+  remote_op_name_to_graph_name = {
+      remote_op_name: graph_name
+  }
+  graph_to_remote_op_input_name_mapping = {
+      graph_name:
+          {remote_op_name:
+              {remote_graph_placeholder_name: parent_graph_input_name}
+          }
+  }
+
+  with beam.Pipeline() as p:
+    # Get the input_pcoll
+    output = input_pcoll | beam_pipeline.ExecuteOneGraph(
+        remote_op_name,
+        remote_op_name_to_graph_name,
+        graph_name_to_specs,
+        graph_to_remote_op_input_name_mapping)
+    # Extract the outputs and store them somewhere.
+  ```
+"""
+
+import copy
+from typing import Dict, Iterator, List, Mapping, Text
+import tensorflow as tf
+import apache_beam as beam
+
+from tfx.experimental.distributed_inference.graphdef_experiments.subgraph_partitioning import execution_spec
+
+
+@beam.ptransform_fn
+@beam.typehints.with_input_types(Dict[Text, Dict[Text, tf.Tensor]])
+@beam.typehints.with_output_types(Dict[Text, Dict[Text, tf.Tensor]])
+def ExecuteGraph(  # pylint: disable=invalid-name
+    pcoll: beam.pvalue.PCollection,
+    remote_op_name: Text,
+    remote_op_name_to_graph_name: Mapping[Text, Text],
+    graph_name_to_specs: Mapping[Text, List[execution_spec.ExecutionSpec]],
+    graph_to_remote_op_input_name_mapping:
+        Mapping[Text, Mapping[Text, Mapping[Text, Text]]]
+) -> beam.pvalue.PCollection:
+  """Executes a graph.
```

Comment:
nit: "A PTransform that executes a graph."

Reply:
Changed.